Update relationship type to take up to 24 bits in high limit format

MishaDemianenko opened the pull request:
Increase relationship type to take up to 24 bits in high limit enterprise record format.
Adapt batch import to support new relationship group types.

changelog: Enhance high limit record format to allow relationship types to take up to 24 bits instead of 16 as it was previously, allowing for many more relationship types than before.

MishaDemianenko committed:
Increase relationship type to take up to 24 bits in high limit enterprise record format.
Update relationship record and relationship group formats.

MishaDemianenko committed:
Adapt batch import to 24 bit relationship group types

burqen commented:
@MishaDemianenko This would be great for the changelog. Please add label and changelog text.

tinwelint commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipGroupCache.java:
```diff
@@ -183,7 +183,8 @@ private long scanForFreeFrom( long startIndex, int groupCount, int type )
             if ( desiredIndex == -1 )
             {
-                int existingType = cache.getShort( candidateIndex, 1 ) & 0xFFFF;
+                // ?? do we need mask here?
```
no we do not

tinwelint commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipGroupCache.java:
```diff
@@ -255,10 +256,11 @@ protected RelationshipGroupRecord fetchNextOrNull()
-                                cache.get6ByteLong( cursor, 1 + 2 ),
-                                cache.get6ByteLong( cursor, 1 + 2 + 6 ),
-                                cache.get6ByteLong( cursor, 1 + 2 + 6 + 6 ),
+                                //do we need mask here?
```
nope :)

tinwelint commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/RelationshipGroupRecordFormat.java:
```diff
@@ -104,8 +106,10 @@ protected void doReadInternal( RelationshipGroupRecord record, PageCursor cursor
         else
         {
+            int typeLowWord = cursor.getShort() & 0xFFFF;
+            int type = (((cursor.getByte() & 0xFF) << Short.SIZE) | typeLowWord);
```
I usually find it safer to first read the `cursor.getByte() & 0xFF` into an `int` variable so that it's guaranteed that `<< Short.SIZE` won't try to left-shift a byte (which obviously means loss of bits). But perhaps just doing the `& 0xFF` converts it to an `int` always. I just know how sensitive these things are sometimes.

MishaDemianenko replied:
that will be int as result, but will extract it to int variable to make it clear and set in int variable :)

tinwelint commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/RelationshipGroupRecordFormat.java:
```diff
@@ -181,7 +188,8 @@ private void readFixedReferencesMethod( RelationshipGroupRecord record, PageCurs
 
-        int type = cursor.getShort() & 0xFFFF;
+        int typeLowWord = cursor.getShort() & 0xFFFF;
+        int type = (((cursor.getByte() & 0xFF) << Short.SIZE) | typeLowWord);
```
would be nice with a utility method for reading a 3-byte int off of a PageCursor.

tinwelint commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/RelationshipGroupRecordFormat.java:
```diff
@@ -218,7 +226,10 @@ private void writeFixedReferencesRecord( RelationshipGroupRecord record, PageCur
-        cursor.putShort( (short) record.getType() );
+        int type = record.getType();
+        cursor.putShort( (short) type );
+        cursor.putByte( (byte) (type >> Short.SIZE) );
```
Also would be nice with a utility for writing a 3-byte int to a PageCursor.

burqen commented:
Friendly reminder to write message for changelog

MishaDemianenko commented:
@burqen it's a prototype so far, but sure as soon as it will be included somewhere it will be updated

tinwelint commented:
@MishaDemianenko ping, decision to go ahead and take this approach the whole way

MishaDemianenko commented:
@tinwelint yep was waiting for migration pr to get in first since it has kind of dependency on how we will migrate

MishaDemianenko committed:
Address PR comments:
 - remove masks
 - introduce int temp variable to be clear
 - extract helper methods

MishaDemianenko committed:
Add additional capability to high limit 3.1 for 3 bytes rel types.
Update way how batch importer retrieve record format.
Update tests

MishaDemianenko commented:
@tinwelint should be ready for review now

tinwelint commented:
Waiting for a migration issue, regarding fixed-references format change, to be sorted out

MishaDemianenko committed:
Update format verification to allow compatible formats of higher generation to be used on top of the stores with more obsolete format.

burqen commented:
@MishaDemianenko Changelog comment is good, but would be more user friendly if it said. "Previously could have x rel types, now you can have y rel types"

MishaDemianenko commented:
Updated mentioning how many bits you had before so it's easy to evaluate
difference

On Thu, Sep 15, 2016 at 10:11 AM, Anton Persson notifications@github.com
wrote:

> @MishaDemianenko https://github.com/MishaDemianenko Changelog comment
> is good, but would be more user friendly if it said. "Previously could have
> x rel types, now you can have y rel types"
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/neo4j/neo4j/pull/7800#issuecomment-247264657, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AMEgk3Uby61BmVu2uvekQ_xHJixx_B_nks5qqP2xgaJpZM4JsfxJ
> .
